### PR TITLE
feat(terraform): Adding yaml based build time policies for corresponding PC run time policies

### PR DESCRIPTION
**Pull request by ssiddardha**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

This PR has 5 YAML - based checks with detailed description below

Checkov Title - Ensure AWS CloudFront distribution is using secure SSL protocols for HTTPS communication

PC Policy ID - 39df6f76-fc34-4660-97a1-fc967e3abe33

PC Policy Title - AWS CloudFront distribution is using insecure SSL protocols for HTTPS communication

Compliance Standard - ACSC Information Security Manual (ISM)-ISM-0109, ACSC Information Security Manual (ISM)-ISM-1228, APRA (CPS 234) Information Security-CPS234-23, APRA (CPS 234) Information Security-CPS234-27, APRA (CPS 234) Information Security-CPS234-34, APRA (CPS 234) Information Security-CPS234-35, APRA (CPS 234) Information Security-CPS234-36, AWS Foundational Security Best Practices standard-Logging, Australian Energy Sector Cyber Security Framework (AESCSF)-SA-1B, Brazilian Data Protection Law (LGPD)-Article 31, Brazilian Data Protection Law (LGPD)-Article 48, CIS AWS 3 Tier Web Architecture Benchmark v.1.0.0-5.3, CIS Controls v7.1-12.5, CIS Controls v7.1-6.2, CIS Controls v8-13.6, CIS Controls v8-8.2, CSA CCM v.4.0.1-LOG-05, CSA CCM v.4.0.1-LOG-06, CSA CCM v.4.0.1-LOG-08, CSA CCM v.4.0.1-LOG-13, CyberSecurity Law of the People's Republic of China-Article 55, CyberSecurity Law of the People's Republic of China-Article 56, Cybersecurity Maturity Model Certification (CMMC) v.1.02-AU.3.046, FFIEC-D2.MA.Ma.B.1, FFIEC-D3.DC.Ev.B.1, Fedramp (Moderate)-RA-05 (08), HITRUST CSF v.9.6.0-09.aa, HITRUST CSF v.9.6.0-09.ab, HITRUST CSF v.9.6.0-09.m, HITRUST v.9.4.2-Control Reference:09.ab, HITRUST v.9.4.2-Control Reference:09.ac, ISO/IEC 27002:2013-12.4.1, ISO/IEC 27002:2013-12.4.3, ISO/IEC 27002:2013-12.4.4, ISO/IEC 27002:2013-16.1.1, ISO/IEC 27017:2015-12.4.1, ISO/IEC 27017:2015-12.4.4, ISO/IEC 27017:2015-16.1.2, MAS TRM 2021-7.5.7, MITRE ATT&CK v10.0-T1562.008 - Impair Defenses:Disable Cloud Logs, MLPS 2.0 (Level 2)-7.1.3.5, MLPS 2.0 (Level 2)-7.1.4.3, MLPS 2.0-8.1.5.4, NIST 800-53 Rev 5-Vulnerability Monitoring and Scanning | Review Historic Audit Logs, NIST 800-53 Rev4-RA-5 (8), NIST CSF-PR.PT-1, NIST SP 800-171 Revision 2-3.3.4, NIST SP 800-172-3.14.2e, New Zealand Information Security Manual (NZISM v3.4)-16.6, PCI DSS v3.2.1-10.2.3, PCI DSS v3.2.1-10.6, PCI DSS v4.0-1.2.1, PCI DSS v4.0-1.2.5, PCI DSS v4.0-1.2.6, PCI DSS v4.0-10.2.1, PCI DSS v4.0-10.2.1.1, PCI DSS v4.0-10.2.1.2, PCI DSS v4.0-10.2.1.3, PCI DSS v4.0-10.2.1.4, PCI DSS v4.0-10.2.1.5, PCI DSS v4.0-10.2.1.6, PCI DSS v4.0-10.2.1.7, PCI DSS v4.0-10.2.2, PCI DSS v4.0-2.2.4, PCI DSS v4.0-5.3.4, PCI DSS v4.0-6.4.1, PCI DSS v4.0-6.4.2, Risk Management in Technology (RMiT)-10.61, Risk Management in Technology (RMiT)-10.66, Secure Controls Framework (SCF) - 2022.2.1-MON-05.1, Secure Controls Framework (SCF) - 2022.2.1-MON-08.2

Remediation Steps -

1. Sign into the AWS console
2. In the console, select the specific region from region drop down on the top right corner, for which the alert is generated
3. Navigate to CloudFront Distributions Dashboard
4. Click on the reported distribution
5. On 'General' tab, Click on 'Edit' button
6. On 'Edit Distribution' page, Set 'Logging' to 'On', choose a 'Bucket for Logs' and 'Log Prefix' as desired
7. Click on 'Yes, Edit'
 
----------------------------------------------------------------------------------------------------------------------------

Checkov Title - Ensure AWS API Gateway endpoints uses client certificate authentication

PC Policy ID - 82493c0b-f4d0-4a82-a5b5-a85df8e78e19

PC Policy Title - AWS API Gateway endpoints without client certificate authentication

Compliance Standard - ACSC Information Security Manual (ISM)-ISM-1603, APRA (CPS 234) Information Security-CPS234-23, APRA (CPS 234) Information Security-CPS234-27, APRA (CPS 234) Information Security-CPS234-34, AWS Foundational Security Best Practices standard-Data protection, Brazilian Data Protection Law (LGPD)-Article 46, Brazilian Data Protection Law (LGPD)-Article 6, CCPA 2018-1798.150(a)(1), CIS Controls v7.1-16.2, CIS Controls v7.1-4.5, CIS Controls v8-12.5, CIS Controls v8-6.6, CSA CCM v.4.0.1-AIS-01, CSA CCM v.4.0.1-AIS-02, CSA CCM v.4.0.1-AIS-04, CSA CCM v.4.0.1-CCC-01, CSA CCM v.4.0.1-GRC-03, CSA CCM v.4.0.1-IVS-04, CSA CCM v.4.0.1-UEM-06, CSA CCM v3.0.1-EKM-01, CSA CCM v3.0.1-IAM-02, CSA CCM v3.0.1-IAM-04, CSA CCM v3.0.1-IAM-08, CSA CCM v3.0.1-IAM-12, CSA CCM v3.0.1-IVS-11, CSA CCM v3.0.1-MOS-16, CyberSecurity Law of the People's Republic of China-Article 21, Cybersecurity Maturity Model Certification (CMMC) v.1.02-AC.2.013, FFIEC-D3.PC.Am.B.6, HITRUST CSF v.9.6.0-01.j, HITRUST CSF v.9.6.0-01.q, HITRUST CSF v9.3-Control Reference:01.q, HITRUST v.9.4.2-Control Reference:10.a, ISO 27001:2013-A.6.2.2, ISO/IEC 27002:2013-12.1.2, ISO/IEC 27002:2013-14.1.1, ISO/IEC 27002:2013-14.1.2, ISO/IEC 27002:2013-14.2.1, ISO/IEC 27002:2013-14.2.2, ISO/IEC 27017:2015-12.1.2, ISO/IEC 27017:2015-14.1.1, ISO/IEC 27017:2015-14.1.2, ISO/IEC 27017:2015-14.2.1, ISO/IEC 27017:2015-14.2.5, ISO/IEC 27017:2015-5.1.1, ISO/IEC 27018:2019-12.1.2, MAS TRM 2021-7.2.1, MAS TRM 2021-7.2.2, MLPS 2.0 (Level 2)-7.1.4.1, MLPS 2.0-8.1.4.1, NIST 800-53 Rev 5-Authenticator Management | Public Key-based Authentication, NIST 800-53 Rev4-IA-5 (2)(a), NIST CSF-PR.IP-1, NIST SP 800-171 Revision 2-3.4.2, NIST SP 800-172-3.4.1e, NIST SP 800-172-3.4.2e, New Zealand Information Security Manual (NZISM v3.4)-17.1, New Zealand Information Security Manual (NZISM v3.4)-22.1, PCI DSS v3.2.1-6.3, PIPEDA-4.7.3, SOC 2-CC6.1, SOC 2-CC8.1, Secure Controls Framework (SCF) - 2022.2.1-IAC-21

Remediation Steps -

These instructions assume you already completed Generate a Client Certificate Using the API Gateway Console. If not please generate a client certificate by following below steps and then Configure an API to Use SSL Certificates.
Steps to Generate a Client Certificate Using the API Gateway Console:
1. Login to AWS Console
2. Go to API Gateway console
3. In the main navigation pane (Left Panel), choose Client Certificates.
4. From the Client Certificates pane, choose Generate Client Certificate.
5. Optionally, for Edit, choose to add a descriptive title for the generated certificate and choose Save to save the description. API Gateway generates a new certificate and returns the new certificate GUID, along with the PEM-encoded public key.

Steps to Configure an API to Use SSL Certificates:
1. Login to AWS Console
2. Go to API Gateway console
3. In the API Gateway console, create or open an API for which you want to use the client certificate. Make sure the API has been deployed to a stage (Left Panel).
4. Choose Stages under the selected API and then choose a stage (Left Panel).
5. In the Stage Editor panel, select a certificate under the Client Certificate section.
6. Click Save Changes

----------------------------------------------------------------------------------------------------------------------------
Checkov Title: Ensure AWS API gateway request is validated

PC Policy ID - fccd881f-26d1-485d-a742-3602afe72035

PC Policy Title - AWS API gateway request parameter is not validated

Compliance Standard - APRA (CPS 234) Information Security-CPS234-23, APRA (CPS 234) Information Security-CPS234-27, APRA (CPS 234) Information Security-CPS234-34, Brazilian Data Protection Law (LGPD)-Article 49, CSA CCM v.4.0.1-AIS-01, CSA CCM v.4.0.1-AIS-02, CSA CCM v.4.0.1-AIS-04, CSA CCM v.4.0.1-CCC-01, CSA CCM v.4.0.1-GRC-03, CSA CCM v.4.0.1-IVS-04, CSA CCM v.4.0.1-UEM-06, Cybersecurity Maturity Model Certification (CMMC) v.1.02-CM.2.062, HITRUST v.9.4.2-Control Reference:10.a, ISO/IEC 27002:2013-12.1.2, ISO/IEC 27002:2013-14.1.1, ISO/IEC 27002:2013-14.1.2, ISO/IEC 27002:2013-14.2.1, ISO/IEC 27002:2013-14.2.2, ISO/IEC 27017:2015-12.1.2, ISO/IEC 27017:2015-14.1.1, ISO/IEC 27017:2015-14.1.2, ISO/IEC 27017:2015-14.2.1, ISO/IEC 27017:2015-14.2.5, ISO/IEC 27017:2015-5.1.1, ISO/IEC 27018:2019-12.1.2, MAS TRM 2021-7.2.1, MAS TRM 2021-7.2.2, MLPS 2.0 (Level 2)-7.1.10.8, NIST CSF-PR.IP-1, NIST SP 800-171 Revision 2-3.4.2, NIST SP 800-172-3.4.1e, NIST SP 800-172-3.4.2e, PCI DSS v3.2.1-6.3, Secure Controls Framework (SCF) - 2022.2.1-MON-06.1

Remediation Steps -

1. Login to AWS management console
2. Navigate to 'API Gateway' service
3. Select the region for which the API gateway is reported.
4. Find the alerted API by the API gateway ID which is the first part of reported resource and click on it
5. Navigate to the reported method
6. Click on the clickable link of 'Method Request'
7. Under section ‘Settings' ,click on the pencil symbol for 'Request Validator' field
8. From the dropdown, Select the type of Request Validator as per the requirement
9. Click on the tick symbol next to it to save the changes

----------------------------------------------------------------------------------------------------------------------------

Checkov Title: Ensure AWS EMR cluster is configured with security configuration

PC Policy ID - fdb20dde-b829-4385-b6b8-eb29f6379e8c

PC Policy Title - AWS EMR cluster is not configured with security configuration

Compliance Standard - APRA (CPS 234) Information Security-CPS234-23, APRA (CPS 234) Information Security-CPS234-27, APRA (CPS 234) Information Security-CPS234-34, Brazilian Data Protection Law (LGPD)-Article 16, Brazilian Data Protection Law (LGPD)-Article 40, CCPA 2018-1798.150(a)(1), CIS Controls v7.1-5.1, CIS Controls v8-4.6, CSA CCM v.4.0.1-AIS-01, CSA CCM v.4.0.1-AIS-02, CSA CCM v.4.0.1-AIS-04, CSA CCM v.4.0.1-CCC-01, CSA CCM v.4.0.1-GRC-03, CSA CCM v.4.0.1-IVS-04, CSA CCM v.4.0.1-UEM-06, CyberSecurity Law of the People's Republic of China-Article 25, Cybersecurity Maturity Model Certification (CMMC) v.1.02-AU.2.042, Fedramp (Low)-CM-06, Fedramp (Moderate)-CM-06, HITRUST CSF v.9.6.0-10.k, HITRUST CSF v.9.6.0-10.m, HITRUST v.9.4.2-Control Reference:10.a, ISO/IEC 27002:2013-12.1.2, ISO/IEC 27002:2013-14.1.1, ISO/IEC 27002:2013-14.1.2, ISO/IEC 27002:2013-14.2.1, ISO/IEC 27002:2013-14.2.2, ISO/IEC 27017:2015-12.1.2, ISO/IEC 27017:2015-14.1.1, ISO/IEC 27017:2015-14.1.2, ISO/IEC 27017:2015-14.2.1, ISO/IEC 27017:2015-14.2.5, ISO/IEC 27017:2015-5.1.1, ISO/IEC 27018:2019-12.1.2, MAS TRM 2021-7.2.1, MAS TRM 2021-7.2.2, MLPS 2.0 (Level 2)-7.1.10.8, MLPS 2.0-8.1.4.4, NIST 800-53 Rev 5-Configuration Settings, NIST 800-53 Rev4-CM-6, NIST CSF-PR.IP-1, NIST SP 800-171 Revision 2-3.4.2, NIST SP 800-172-3.4.1e, NIST SP 800-172-3.4.2e, PCI DSS v3.2.1-6.3, PIPEDA-4.7.3, Secure Controls Framework (SCF) - 2022.2.1-NET-04.10

Remediation Steps -

1. Log in to the AWS Console
2. In the console, select the specific region from region drop down on the top right corner, for which the alert is generated.
3. Navigate to 'EMR' dashboard from 'Services' dropdown
4. Go to 'Security configurations', click 'Create'
5. On the Create security configuration window,
6. In 'Name' box, provide a name for the new EMR security configuration
7. Follow below link to configure a security configuration
https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-create-security-configuration.html
8. Click on 'Create' button
9. On the left menu of EMR dashboard Click 'Clusters'
10. Select the EMR cluster for which the alert has been generated and click on the 'Clone' button from the top menu.
11. In the Cloning popup, choose 'Yes' and Click 'Clone'
12. On the Create Cluster page, in the Security Options section, click on 'security configuration'
13. From the 'Security configuration' drop down select the name of the security configuration created at step 4 to step 8, click 'Create Cluster'.
14. Once you the new cluster is set up verify its working and terminate the source cluster in order to stop incurring charges for it
15. On the left menu of EMR dashboard Click 'Clusters', from the list of clusters select the source cluster which is alerted
16. Click on the 'Terminate' button from the top menu
17. On the 'Terminate clusters' pop-up, click 'Terminate'

----------------------------------------------------------------------------------------------------------------------------

Checkov Title - Ensure AWS ElasticSearch Fine-grained access control is enabled

PC Policy ID - 45987be9-259e-4940-ae03-351c68b8d3d8

PC Policy Title - AWS OpenSearch Fine-grained access control is disabled

Compliance Standard - ACSC Information Security Manual (ISM)-ISM-1566, CIS Controls v7.1-16.9, CIS Controls v8-3.3, CIS Controls v8-5.4, Cybersecurity Maturity Model Certification (CMMC) v.2.0 (Level 1)-AC.L1-3.1.1, Cybersecurity Maturity Model Certification (CMMC) v.2.0 (Level 1)-AC.L1-3.1.2, FFIEC-D3.PC.Am.B.3, HITRUST CSF v.9.6.0-01.e, MLPS 2.0 (Level 2)-7.1.4.2, PCI DSS v4.0-1.3.1, Secure Controls Framework (SCF) - 2022.2.1-IAC-21

Remediation Steps -

Refer the following URL for configuring Fine-grained access control on your AWS OpenSearch:
https://docs.aws.amazon.com/opensearch-service/latest/developerguide/fgac.html#fgac-forget

Notes: 
1. You can't enable fine-grained access control on existing domains, only new ones. After you enable fine-grained access control, you can't disable it.
2. Fine-grained access control is supported only from ElasticSearch 6.7 or later. To upgrade older versions of AWS OpenSearch please refer to the URL given below,
https://docs.aws.amazon.com/opensearch-service/latest/developerguide/version-migration.html

### CODE REVIEW

File: checkov/terraform/checks/graph_checks/aws/APIGatewayEndpointsUsesCertificateForAuthentication.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+metadata:
+  id: "CKV2_AWS_51"
+  name: "Ensure AWS API Gateway endpoints uses client certificate authentication"
+  category: "GENERAL_SECURITY"
+definition:
+  cond_type: "attribute"
+  resource_types:
+  - "aws_api_gateway_stage"
+  - "aws_apigatewayv2_stage"
```
Comment (gruebel) [inline checkov/terraform/checks/graph_checks/aws/APIGatewayEndpointsUsesCertificateForAuthentication.yaml:9]: for API Gateway V2 you can only set `client_certificate_id`, if the actual API protocol type is websocket. So you would need to check for the connection to an `aws_apigatewayv2_api` and its `protocol_type`

File: checkov/terraform/checks/graph_checks/aws/OpenSearchDomainHasFineGrainedControl.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+metadata:
+  id: "CKV2_AWS_52"
+  name: "Ensure AWS ElasticSearch Fine-grained access control is enabled"
```
Comment (gruebel) [inline checkov/terraform/checks/graph_checks/aws/OpenSearchDomainHasFineGrainedControl.yaml:3]:
```suggestion
  name: "Ensure AWS ElasticSearch/OpenSearch Fine-grained access control is enabled"
```

File: tests/terraform/graph/checks/resources/EMRClusterHasSecurityConfiguration/main.tf [hunk 3]
```diff
@@ -0,0 +1,59 @@
+resource "aws_emr_cluster" "pass" {
+  # ... other configuration ...
+
+  # EMR version must be 5.23.0 or later
+  release_label = "emr-5.24.1"
+
+  security_configuration = "securityGroupId"
```
Comment (gruebel) [inline tests/terraform/graph/checks/resources/EMRClusterHasSecurityConfiguration/main.tf:7]:
```suggestion
  security_configuration = "example"
```
not so important, but it is definitely not a security group 😄

**Review by gruebel — COMMENTED** [review]
nice work 🍻 added some comments

File: checkov/terraform/checks/graph_checks/aws/APIGatewayEndpointsUsesCertificateForAuthentication.yaml (hunk 1 above)
Comment (ssiddardha) [reply checkov/terraform/checks/graph_checks/aws/APIGatewayEndpointsUsesCertificateForAuthentication.yaml:9]: thanks for pointing it out, I missed it in terraform documentation. I will update the yaml check accordingly.

Comment (ssiddardha) [reply checkov/terraform/checks/graph_checks/aws/APIGatewayEndpointsUsesCertificateForAuthentication.yaml]: @gruebel Updated check according to the comment and updated pass fail test cases.

**Review by gruebel — APPROVED** [review]
great work 🍻

**Review by rotemavni — APPROVED** [review]

**Pull request merged**